pkg/gitlab/groupreadwriter.go

```diff
@@ -0,0 +1,358 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitlab
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	gitlab "github.com/xanzy/go-gitlab"
+
+	"github.com/abcxyz/pkg/cache"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/pkg/sets"
+	"github.com/abcxyz/team-link/pkg/groupsync"
+	"github.com/abcxyz/team-link/pkg/utils"
+)
+
+const (
+	// DefaultCacheDuration is the default time to live for the user and group caches.
+	// We don't expect user info (e.g. username etc.) nor group info (group name etc.)
+	// to change frequently so a time to live of 1 day is the default.
+	DefaultCacheDuration = time.Hour * 24
+)
+
+type Config struct {
+	includeSubGroups bool
+	cacheDuration    time.Duration
+}
+
+type Opt func(writer *Config)
+
+// WithCacheDuration set the time to live for the user and group cache entries.
+func WithCacheDuration(duration time.Duration) Opt {
+	return func(config *Config) {
+		config.cacheDuration = duration
+	}
+}
+
+// WithoutSubGroupsAsMembers toggles off treating subgroups as members of their parent group.
+// When this option is used GroupReadWriter.GetMembers will only return user members of the group.
+// Similarly, GroupReadWriter.SetMembers will only consider user members when setting members.
+func WithoutSubGroupsAsMembers() Opt {
+	return func(config *Config) {
+		config.includeSubGroups = false
+	}
+}
+
+type GroupReadWriter struct {
+	client           *gitlab.Client
+	userCache        *cache.Cache[*gitlab.User]
+	groupCache       *cache.Cache[*gitlab.Group]
+	includeSubGroups bool
+}
+
+func NewGroupReadWriter(client *gitlab.Client, opts ...Opt) *GroupReadWriter {
+	config := &Config{
+		includeSubGroups: true,
+		cacheDuration:    DefaultCacheDuration,
+	}
+
+	for _, opt := range opts {
+		opt(config)
+	}
+	return &GroupReadWriter{
+		client:           client,
+		userCache:        cache.New[*gitlab.User](config.cacheDuration),
+		groupCache:       cache.New[*gitlab.Group](config.cacheDuration),
+		includeSubGroups: config.includeSubGroups,
+	}
+}
+
+// GetUser retrieves the GitLab user with the given ID. The ID is the GitLab user's login.
+func (rw *GroupReadWriter) GetUser(ctx context.Context, userID string) (*groupsync.User, error) {
+	user, err := rw.getGitLabUser(ctx, userID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get user: %w", err)
+	}
+	return &groupsync.User{
+		ID:         user.Username,
+		Attributes: user,
+	}, nil
+}
+
+func (rw *GroupReadWriter) getGitLabUser(ctx context.Context, userID string) (*gitlab.User, error) {
+	if user, ok := rw.userCache.Lookup(userID); ok {
```

Comment:
All of these functions that check the cache are racey because there's no lock on the cache lookup. Use WriteThruLookup instead? https://github.com/abcxyz/pkg/blob/main/cache/cache.go#L155

Reply:
Done